### PR TITLE
Cache mods

### PR DESCRIPTION
**Pull request by TOLoneWolf**
To speed up installing the same versions:
- Check if vanilla version was downloaded, skipping asset download if it was.
- Check if forge version was already downloaded and patched, skipping asset download and patch if it was.

- Added mod caching:
  - ~~- Toggle switch for caching.~~
  - Cache on installing individual mods from mod browser.
  - ~~Cache mods download with mod pack manifest.~~

!! Warning: Requires new instance meta data !!
- Added "categorySection.path". This is required for correct destination of curseforge file (mod, resourcepack, world save). This fixes modpacks containing resource packs.

Example:
Modpack with resource pack for testing changing destination of downloaded curseforge files.
HR: New Beginnings version: New Beginnings-1.2.1.zip
https://www.curseforge.com/minecraft/modpacks/hr-new-beginnings/files/2908388

texturepack(resourcepack): Fancy Skies

### CODE REVIEW

**Comment by killpowa** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/common/reducers/actions.js  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/gorilla-devs/GDLauncher/pullRequest?prid=5382015&bid=17646659)

**Comment by TOLoneWolf** [comment]
#424

**Comment by blarfoon** [comment]
This is a cool idea, but it's implemented in an unconvenient way.
The current way this is implemented creates a lot of duplication.
What we want to do, is use the mods already present in other instances and just copy them over if they are present.

**Comment by TOLoneWolf** [comment]
Yeah was planning on cleaning this up a bit and giving it instance only cache  option and full cache like it is now.

**Comment by blarfoon** [comment]
I think it's better to just use instance caching. I don't see any benefit on using the current method. Yeah maybe you can keep some more mods cached and skip some downloads, but at what storage cost.

**Comment by killpowa** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/common/reducers/actions.js  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/gorilla-devs/GDLauncher/pullRequest?prid=5382015&bid=17646659)

**Comment by killpowa** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/common/reducers/actions.js  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/gorilla-devs/GDLauncher/pullRequest?prid=5382015&bid=17646659)

File: src/common/modals/Settings/components/General.js [hunk 1]
```diff
@@ -392,6 +413,127 @@ const General = () => {
...
+          Cache Mods Using Dedicated&nbsp; <FontAwesomeIcon icon={faHdd} />
+        </Title>
+        <DiscordRpc>
+          <p
+            css={`
+              width: 500px;
+            `}
+          >
+            Enable / disable - Caching mods from curseforge to a dedicated cache
+            folder. Keeps a copy of every mod installed to a special cache folder.
```
Comment (killpowa) [inline src/common/modals/Settings/components/General.js:486]: Codacy Issue found: [Insert `⏎···········`](https://app.codacy.com/gh/gorilla-devs/GDLauncher/pullRequest?prid=5382015)

**Comment by killpowa** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/common/reducers/actions.js  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/gorilla-devs/GDLauncher/pullRequest?prid=5382015&bid=17646659)

**Comment by killpowa** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/common/reducers/actions.js  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/gorilla-devs/GDLauncher/pullRequest?prid=5382015&bid=17646659)

**Comment by TOLoneWolf** [comment]
Modpack with resource pack for testing changing destination of downloaded curseforge files.
HR: New Beginnings version: New Beginnings-1.2.1.zip 
https://www.curseforge.com/minecraft/modpacks/hr-new-beginnings/files/2908388

texturepack(resourcepack): Fancy Skies

**Comment by killpowa** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/common/reducers/actions.js  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/gorilla-devs/GDLauncher/pullRequest?prid=5382015&bid=17646659)

**Comment by killpowa** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/common/reducers/actions.js  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/gorilla-devs/GDLauncher/pullRequest?prid=5382015&bid=17646659)

**Comment by blarfoon** [comment]
I think we should just remove the dedicated folder caching. It makes everything just overcomplicated for no reason. I don't see the point of having a local mirror of curseforge basically

**Comment by blarfoon** [comment]
What I'm saying is that there should be no option in the settings. It should "just work".
The user installs a new modpack/mod and we try to fetch mods from other instances, the user doesn't need to know

**Comment by DylanBeMe** [comment]
Have a list of currently downloaded mods and where they are and fetch from other installed packs?

**Comment by TOLoneWolf** [comment]
Correct the switches are not really needed (as everything should work right in instances installation). Added them just in case something breaks and it needs to be disabled for debugging.

As for dedicated cache I find it use full but that's just my point and my use case. But I guess if that's what you want I can remove it.

> Have a list of currently downloaded mods and where they are and fetch from other installed packs?
@DylanHameleers 

Yep that is what this does bud.

**Comment by AzureDoom** [comment]
> Check if vanilla version was downloaded, skipping asset download if it was.
> 
> Check if forge version was already downloaded and patched, skipping asset download and patch if it was.

I suggest these say should say, as it would make reinstalling a modpack or doing a vanilla download, even faster.

**Comment by TOLoneWolf** [comment]
those are staying just the switches to disabled them is going away i guess.

image

**Comment by killpowa** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/common/reducers/actions.js  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/gorilla-devs/GDLauncher/pullRequest?prid=5382015&bid=17646659)

**Comment by killpowa** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/common/reducers/actions.js  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/gorilla-devs/GDLauncher/pullRequest?prid=5382015&bid=17646659)

**Comment by TOLoneWolf** [comment]
There dedicated cache stuff removed.
Switches removed.

**Comment by blarfoon** [comment]
I took a look at the code. I feel like the `.complete` thing is unnecessary, we just need to ensure we save the `.json` at the end of the procedures and then we can use that directly to eventually skip

**Comment by blarfoon** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/common/reducers/actions.js  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/gorilla-devs/GDLauncher/pullRequest?prid=5382015&bid=17646659)

**Pull request closed without merging**